smithy-aws-traits/src/main/java/software/amazon/smithy/aws/traits/ErrorRenameValidator.java

```diff
@@ -0,0 +1,96 @@
+ * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.aws.traits;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import software.amazon.smithy.aws.traits.protocols.AwsJson1_1Trait;
+import software.amazon.smithy.aws.traits.protocols.AwsProtocolTrait;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.traits.ErrorTrait;
+import software.amazon.smithy.model.validation.AbstractValidator;
+import software.amazon.smithy.model.validation.ValidationEvent;
+import software.amazon.smithy.utils.SetUtils;
+import software.amazon.smithy.utils.SmithyInternalApi;
+
+@SmithyInternalApi
+public final class ErrorRenameValidator extends AbstractValidator {
+
+    /**
+     * By default, server-side implementation of AWS Protocols send the error shape ID on the wire,
+     * and clients can use it to resolve the error type.
+     *
+     * However, some protocols implementations, including AWS JSON 1.1, only send the shape name.
+     * If there are conflicting shape IDs e.g. smithy.service#ServiceError and smithy.other#ServiceError,
+     * clients are unable to resolve the error type with the shape name alone.
+     *
+     * In addition, Server-side implementation of these protocols don't handle or send renamed shape names.
+     *
+     * Hence, error shape renaming are not supported for these protocols.
+     */
+    private static final Set<Class<? extends AwsProtocolTrait>> UNSUPPORTED_PROTOCOLS =
+        SetUtils.of(AwsJson1_1Trait.class);
+
+    @Override
+    public List<ValidationEvent> validate(Model model) {
+        List<ValidationEvent> events = new ArrayList<>();
+        for (ServiceShape shape : model.getServiceShapes()) {
+            validate(model, shape, events);
+        }
+        return events;
+    }
+
+    private void validate(Model model, ServiceShape service, List<ValidationEvent> events) {
+        final Map<ShapeId, String> renames = service.getRename();
+
+        if (renames.isEmpty()) {
+            return;
+        }
+
+        Set<String> unSupportedProtocols = getUnsupportedProtocols(service);
+
+        if (unSupportedProtocols.isEmpty()) {
+            return;
+        }
+
+        renames.keySet().stream()
```

Comment:
The use of streams here seems unnecessary, especially given the use of forEach to modify the passed in List.